File: projects/M21C_ls/examples/AMS_24/IVs/Plot_script_sample.m
clear
data_path = '/discover/nobackup/qliu/for_Andy/';

% choose 2 data for comparison
D1_version = 'OLv7_M36_FPprcp';
D2_version = 'OLv7_M36';

% preprocessed skill files
f_D1 =  [data_path,'SMPL3_',D1_version,'_IVD_IVS_stats_lag2day_201504_202103.mat'];
f_D2 =  [data_path,'SMPL3_',D2_version,'_IVD_IVS_stats_lag2day_201504_202103.mat'];

% load skill values, note it's R^2
D1 = load(f_D1,'R2_ivs_obs','R2_ivs_mod');
D2 = load(f_D2,'R2_ivs_obs','R2_ivs_mod');

% get EASE grid coord info
[lat,lon] = smapeasev2_ind2latlon([0:405],[0:963],'M36');
lon_EASE = repmat(lon',[1,length(lat)]);
lat_EASE = repmat(lat,[length(lon),1]);

% Compute R and exclude points where estimated R's are too poo
R_D1 = sqrt(D1.R2_ivs_mod);  R_D1(R_D1 < 0.1) = NaN;
R_D2 = sqrt(D2.R2_ivs_mod);  R_D2(R_D2 < 0.1) = NaN;
R_OBS = sqrt(D2.R2_ivs_obs); R_OBS(R_OBS < 0.1) = NaN;

R_D1(isnan(R_OBS)) = NaN;
R_D2(isnan(R_OBS)) = NaN;

% Usually plot the differences only as the R values tend to be overestimated with IVs method
% Also, data are stored in 1-D and need to reshape in 2-D
Rdiff_D2minusD1 = reshape(R_D2 - R_D1, 964, 406);

% Plot with own preferences
figure, pcolor(lon_EASE,lat_EASE, Rdiff_D2minusD1); shading('flat') 

File: projects/matlab2python/scripts/plot_script_sample.py
from __future__ import annotations

import argparse
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from scipy.io import loadmat, savemat

# Make the shared EASE grid utilities available
REPO_ROOT = Path(__file__).resolve().parents[1]
sys.path.append(str(REPO_ROOT / "shared" / "python"))
from EASEv2 import EASEv2_ind2latlon  # noqa: E402


def _as_vector(mat_array: np.ndarray) -> np.ndarray:
    """Flatten MATLAB-loaded arrays using column-major order."""
    return np.asarray(mat_array, dtype=float).ravel(order="F")


def load_skill_file(path: Path) -> tuple[np.ndarray, np.ndarray]:
    """Load R2 fields from a MATLAB stats file."""
    data = loadmat(path, simplify_cells=True)
    return _as_vector(data["R2_ivs_obs"]), _as_vector(data["R2_ivs_mod"])


def build_ease_grid() -> tuple[np.ndarray, np.ndarray]:
    """Build latitude/longitude grids for the M36 EASEv2 layout."""
    rows = np.arange(406)
    cols = np.arange(964)
    lat, lon = EASEv2_ind2latlon(rows, cols, "M36")
    lon_grid = np.tile(lon[:, None], (1, lat.size))
    lat_grid = np.tile(lat, (lon.size, 1))
    return lat_grid, lon_grid


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Compute R differences between two IVS skill files and save a .mat output.",
    )
    parser.add_argument(
        "--data-path",
        type=Path,
        default=Path.home() / "for_Andy/IVs",
        help="Directory containing the IVS stats .mat files.",
    )
    parser.add_argument(
        "--d1-version",
        default="OLv7_M36_MULTI_type_13_comb_fp_scaled",
        help="Label for the first dataset (D1).",
    )
    parser.add_argument(
        "--d2-version",
        default="DAv7_M36_ASCAT_type_13_comb_fp_scaled",
        help="Label for the second dataset (D2).",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path.cwd(),
        help="Directory to write the output .mat file.",
    )
    parser.add_argument(
        "--plot",
        action=argparse.BooleanOptionalAction,
        default=True,
        help="Whether to display a pcolor plot of the R difference.",
    )
    args = parser.parse_args()

    data_path = args.data_path.expanduser()
    f_D1 = data_path / f"SMPL3_{args.d1_version}_IVD_IVS_stats_lag2day_201504_202103.mat"
    f_D2 = data_path / f"SMPL3_{args.d2_version}_IVD_IVS_stats_lag2day_201504_202103.mat"

    R2_obs_D1, R2_mod_D1 = load_skill_file(f_D1)
    R2_obs_D2, R2_mod_D2 = load_skill_file(f_D2)

    R_D1 = np.sqrt(R2_mod_D1)
    R_D2 = np.sqrt(R2_mod_D2)
    R_OBS = np.sqrt(R2_obs_D2)

    R_D1[R_D1 < 0.1] = np.nan
    R_D2[R_D2 < 0.1] = np.nan
    R_OBS[R_OBS < 0.1] = np.nan

    R_D1[np.isnan(R_OBS)] = np.nan
    R_D2[np.isnan(R_OBS)] = np.nan

    Rdiff_vector = R_D2 - R_D1

    lat_grid, lon_grid = build_ease_grid()
    Rdiff_grid = np.reshape(Rdiff_vector, lon_grid.shape, order="F")

    if args.plot:
        fig, ax = plt.subplots(figsize=(12, 5))
        mesh = ax.pcolormesh(lon_grid, lat_grid, Rdiff_grid, shading="flat")
        ax.set_xlabel("Longitude")
        ax.set_ylabel("Latitude")
        ax.set_title(f"R diff: {args.d2_version} minus {args.d1_version}")
        fig.colorbar(mesh, ax=ax, label="R difference (D2 - D1)")
        plt.tight_layout()
        plt.show()

    outname = f"Rdiff_{args.d2_version}_minus_{args.d1_version}.mat"
    args.output_dir.mkdir(parents=True, exist_ok=True)
    savemat(
        args.output_dir / outname,
        {
            "Rdiff_vector": Rdiff_vector,
            "lats": lat_grid.reshape(-1, order="F"),
            "lons": lon_grid.reshape(-1, order="F"),
        },
    )


if __name__ == "__main__":
    main()
